### ISSUE TITLE
Inference(?) bug on 0.6

### ISSUE THREAD

**Issue by simonster** [JuliaLang/julia#23170]
```julia
f(::Type{T}) where {T} = "good"
f(::Type{T}) where {T<:DataType} = "oops!"
g(T::DataType) = f(T)

julia> f(Int)
"good"

julia> g(Int)
"oops!"

julia> code_typed(g, Tuple{Type{Int}})
1-element Array{Any,1}:
 CodeInfo(:(begin 
        return $(QuoteNode("good"))
    end))=>String

julia> code_typed(g, Tuple{DataType})
1-element Array{Any,1}:
 CodeInfo(:(begin 
        return "oops!"
    end))=>String

julia> Tuple{Type{Int}} <: Tuple{DataType}
true
```

The behavior on master is correct.

**Comment by tkelman** [comment]
bisect identifies f0e3678d4cfa922341e26ceaf85dba5ce99f0c14 as the fix, from #22701 which is already marked for backporting so should get into 0.6.1 if it doesn't break anything

**Comment by JeffBezanson** [comment]
Closing as fixed.
